<dl align="center">
<h1><br><a href=""><img src="paper/img/misc/groot-logo.png" alt="groot" width="250"></a><br>G R O O T</h1>
<h3><a style="color:#FF6600">G</a>raphing <a style="color:#FF6600">R</a>esistance genes <a style="color:#FF6600">O</a>ut <a style="color:#FF6600">O</a>f me<a style="color:#FF6600">T</a>agenomes</h3>
<p>
<a href="https://travis-ci.org/will-rowe/groot"><img src="https://travis-ci.org/will-rowe/groot.svg?branch=master" alt="travis"></a>
<a href=""><img src="https://img.shields.io/badge/stability-stable-green.svg" alt="status"></a>
<a href="https://will-rowe.gitbooks.io/groot"><img src="https://img.shields.io/badge/docs-passing-green.svg" alt="Documentation Status"></a>
</p>
</dl>

***

## Overview

`GROOT` is a tool to type **Antibiotic Resistance Genes** (**ARGs**) in metagenomic samples. It combines variation graph representation of gene sets with an LSH indexing scheme to allow for fast classification of metagenomic reads. Subsequent hierarchical local alignment of classified reads against graph traversals facilitates accurate reconstruction of full-length gene sequences using a simple scoring scheme.

`GROOT` will output an ARG alignment file (in BAM format) that contains the graph traversals possible for each query read; the alignment file is then used to generate a simple ARG typing report.


## Installation

Check out the [releases](https://github.com/will-rowe/groot/releases) to download a binary. Alternatively, compile the software from source:

`GROOT` is written in Go (v1.9) - to compile from source you will first need the [Go tool chain](https://golang.org/doc/install). Once you have it, try something like this to compile:

```bash
# Clone this repository
git clone https://github.com/will-rowe/groot.git

# Go into the repository and get the package dependencies
cd groot
go get -d -t -v ./...

# Compile the program
go build ./

# Call the program
./groot --help
```


## Quick Start

`GROOT` is called by typing **groot**, followed by the command you wish to run. There are three main commands, **index**, **align** and **report**. This quick start will cover how to get things running but it is recommended to follow the [documentation](https://will-rowe.gitbooks.io/groot) for more advanced use cases.

### 1. Index

Indexing requires a clustered set of reference ARGs (some pre-clustered sets can be found in the `paper/data` directory within this repo).

```bash
groot index -i paper/data/clustered-ARG-databases/arg-annot-90 -o groot-index
```

> Indexing only needs to run once, unless you wish to alter the seeding parameters used during the alignment or the read length has changed.

### 2. Align

Alignment uses the previously generated index files and requires a file of FASTQ reads as input.

```bash
groot align -i groot-index -f paper/data/test-reads/full-argannot-perfect-reads-small.fq.gz > out.bam
```

### 3. Report

Reporting takes the alignment generated by **groot align** and creates an ARG typing report.

```bash
groot report -i out.bam > report.txt
```


## Further Information and citing

Please read the [gitbook](https://will-rowe.gitbooks.io/groot/) for more extensive documentation.

The paper describing `GROOT` is in preparation and will be pre-printed soon.
